brainlife/ezbids
# Add login functionality to ezBIDS

## This PR aims to connect our login implementation for brainlife to ezBIDS.
_NOTE: This PR is also connected to changes to [generate_keys.sh](https://github.com/brainlife/brainlife/pull/229/files)._

### HOW TO TEST

1. Go to chrome and enter this flag in the URL: chrome://flags/#block-insecure-private-network-requests . Set it to disabled
&nbsp;&nbsp;&nbsp;&nbsp;- This flag exists due to security vulnerabilities so make sure to enable again after testing
2. Go to http://149.165.152.76:8080/ezbids/ 
3. Click on Get Started and log in
&nbsp;&nbsp;&nbsp;&nbsp;- ADMIN USER: email: admin@dev.brainlife.io password: 30q49th
&nbsp;&nbsp;&nbsp;&nbsp;- GUEST USER: email: guest@dev.brainlife.io password: 912r3iub
4. Check that you can go to warehouse projects and are logged in already
5. create an ezBIDS session by uploading files
&nbsp;&nbsp;&nbsp;&nbsp;- login as another user in private browsing (you can create a user if you want or use the one you are not logged in with) and check that you cannot access the session
&nbsp;&nbsp;&nbsp;&nbsp;- in the original window, click the Share Session button to open the manage users dialog and add a user.
&nbsp;&nbsp;&nbsp;&nbsp;- use incognito mode and log in as the user you just added. Check that you can access the given session. This user should not be able to see the Share Session button.

### SUMMARY OF CHANGES
1. Removed a bunch of auto generated .js files
2. Added implementation for validating JWTs in auth.ts.
&nbsp;&nbsp;&nbsp;&nbsp;- ezBIDS now issues its own JWTs from the `/download/:session_id/token` which will only be valid for 10 seconds.
&nbsp;&nbsp;&nbsp;&nbsp;- The reason for this is because the `/download/:session_id/*` route often is used in a way that does not allow us to attach HTTP headers. We therefore send a JWT in the URL (albeit a short lived one).
&nbsp;&nbsp;&nbsp;&nbsp;- The `/download/:session_id/*` route is the only valid route that expects and consumes these JWTs. Other routes will expect JWTs in the header.
3. All API routes now are protected except for the healthcheck
7. dev.sh has been modified to be clearer on failure
8. A ManageUsersDialog has been added in order to add/remove users to the session to share it with them. It can only be shared with other brainlife users
9. Note: We use the same JWTs received from the auth service as warehouse. This means that if someone logs into warehouse, they can access ezBIDS and vice versa.
10. A bunch of `<a ...><img ... /></a>` HTML has been replaced with a custom `<AsyncImageLink />`. This is because of the previously mentioned token change.
&nbsp;&nbsp;&nbsp;&nbsp;- previously these elements were bound to the path directly, i.e. `<img :src="/route/to/backend.png" />.` This doesn't work anymore so this case needs to be handled.
11. I've replaced all fetch requests with axios. Axios allows us to add a global http interceptor, which I've defined in `axios.instance.ts`. This attaches a JWT to every outgoing request.
12. I've added route protections to prevent people from navigating to /ezbids/convert unless they are authenticated
13. The manage session button only appears if (1) there is a session and (2) the owner of the session is the current user. 
 &nbsp;&nbsp;&nbsp;&nbsp;- people added to the session cannot add more users, only the owner can
